### PR TITLE
feat: files (mfs) api

### LINKED ISSUE #53: How to add files to MFS?
How can we add files to the MFS using this API?

### PR DESCRIPTION
**Pull request by hacdias**
- Implementation: https://github.com/ipfs/go-ipfs/pull/6786
- Closes #53

This adds support for Files API.

### CODE REVIEW

**Comment by hacdias** [comment]
@Stebalien oh, this is a WIP, I haven't yet added all methods I'm thinking of adding 😄 nor the comments! I will do that and ping you again!

**Comment by hacdias** [comment]
@Stebalien I actually implemented almost everything from the API as far as I know! Please take a look!

File: files.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	Local          bool   `json:",omitempty"`
+	SizeLocal      uint64 `json:",omitempty"`
+}
+
+// FilesAPI specifies an interface to interact with the Mutable File System
+// layer.
+type FilesAPI interface {
+	Copy(ctx context.Context, src, dst string, opts *FilesCopyOptions) error
+	Move(ctx context.Context, src, dst string, opts *FilesMoveOptions) error
+	List(ctx context.Context, path string, opts *FilesListOptions) ([]mfs.NodeListing, error)
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply files.go]: Let's re-use the return type from `Unixfs().Ls()`. I'd rather not expose internal mfs types in this API.

File: files.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	Flush       bool
+	CidBuilder  cid.Builder
+}
+
+type FilesChangeCidOptions struct {
+	Flush      bool
+	CidBuilder cid.Builder
+}
+
+type FileInfo struct {
```
Comment (Stebalien) [inline files.go]: docs

File: files.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	Hash           string
+	Size           uint64
+	CumulativeSize uint64
+	Blocks         int
+	Type           string
```
Comment (Stebalien) [inline files.go]: Please use the file types from `unixfs.go`.

File: files.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	CidBuilder  cid.Builder
+}
+
+type FilesChangeCidOptions struct {
+	Flush      bool
+	CidBuilder cid.Builder
+}
+
+type FileInfo struct {
+	Hash           string
```
Comment (Stebalien) [inline files.go]: Let's use `Cid cid.Cid`.

File: files.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	Size           uint64
+	CumulativeSize uint64
+	Blocks         int
+	Type           string
+	WithLocality   bool   `json:",omitempty"`
+	Local          bool   `json:",omitempty"`
+	SizeLocal      uint64 `json:",omitempty"`
```
Comment (Stebalien) [inline files.go]: Hm. I remember not liking this when it first came up. Can we come up with a better (internal) API and translate back at the edges? I'll think about this a bit but proposals are very welcome. For historical context: https://github.com/ipfs/go-ipfs/pull/4638

File: files.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+type FilesChangeCidOptions struct {
+	Flush      bool
+	CidBuilder cid.Builder
+}
+
+type FileInfo struct {
+	Hash           string
+	Size           uint64
+	CumulativeSize uint64
+	Blocks         int
```
Comment (Stebalien) [inline files.go]: I have no idea why we called this "blocks". Let's call it what it is, Children (NumChildren?).

File: files.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"context"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-cidutil/cidenc"
+	"github.com/ipfs/go-mfs"
+)
+
+type FilesCopyOptions struct {
```
Comment (Stebalien) [inline files.go]: Docs. Lots of docs.

**Comment by hacdias** [comment]
@Stebalien before continuing, I think it's best to define which functions to keep 😄 You said `Move`, `Rename` (one of these should be copy I presume) and `Remove`.

What if we keep:

1. Copy
2. Move or Rename (which name is better?)
3. List
4. Remove
5. Stat
6. Make Dir

With that, we would remove `Flush` (all operations would simply flush), `Read`  (we get Stat and then get the Cid and use Unixfs API for now) and `Write` (Unixfs + Copy).

What do you think?

**Comment by Stebalien** [comment]
Status update: I believe we paused this in favor of trying to revive https://github.com/ipfs/interface-go-ipfs-core/pull/19.

Closing for now.

**Pull request closed without merging**